httpgrpc Server can now use message from HTTP header when converting HTTP response to error

pstibrany opened the pull request:
**What this PR does**:

httpgrpc Server (implementation of `httpgrpc.HTTP` gRPC service) converts HTTP responses to errors under certain conditions. In this conversion it uses HTTP response body as an error message. If HTTP response body is not a valid UTF-8 string, then such converted error will not be protobuf-marshalable anymore.

This PR adds option to supply alternative error message for such conversion by setting special HTTP header, that is recognized by httpgrpc Server.

**Which issue(s) this PR fixes**:

Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

aknuds1 commented on httpgrpc/server/server_test.go:
```diff
@@ -326,3 +327,125 @@ func TestTracePropagation(t *testing.T) {
+			s := NewServer(h)
+			req := &httpgrpc.HTTPRequest{Method: "GET", Url: "/test"}
+			resp, err := s.Handle(context.Background(), req)
+			require.NotNil(t, err)
```
[Nit]
```suggestion
			require.Error(t, err)
```

aknuds1 commented on httpgrpc/server/server_test.go:
```diff
@@ -326,3 +327,125 @@ func TestTracePropagation(t *testing.T) {
+	}
+}
+
+func Test(t *testing.T) {
```
Is this copy-pasted from `TestGrpcErrorsHaveCorrectMessage` by mistake?

pstibrany replied:
lol, yes. no idea how that happened :) thanks for spotting!

aknuds1 commented on httpgrpc/server/server_test.go:
```diff
@@ -326,3 +327,125 @@ func TestTracePropagation(t *testing.T) {
+		require.False(t, IsHandledByHttpgrpcServer(context.Background()))
+	})
+
+	testHeader := "X-HandledByHttpgrpcServer"
```
[Nit]
```suggestion
	const testHeader = "X-HandledByHttpgrpcServer"
```

aknuds1 commented on httpgrpc/server/server_test.go:
```diff
@@ -326,3 +327,125 @@ func TestTracePropagation(t *testing.T) {
+	t.Run("handler runs from httpgrpc.Server", func(t *testing.T) {
+		s := NewServer(handler)
+		resp, err := s.Handle(context.Background(), &httpgrpc.HTTPRequest{Method: "GET", Url: "/test"})
+		require.Nil(t, err)
```
[Nit]
```suggestion
		require.NoError(t, err)
```

aknuds1 reviewed (approved):
LGTM, modulo nits.

pstibrany commented:
> LGTM, modulo nits plus what I think is a left behind copy-paste.

Thanks for review, I've addressed your feedback in https://github.com/grafana/dskit/pull/531/commits/4330714569d0af9eb281a6411190af1ca3409088
